PIMbench/cpp-knn/baselines/GPU/knncuda.cu

```diff
@@ -0,0 +1,438 @@
+        int max_count = 0;
+        for (int i = 0; i < k; ++i) {
+            if (vote_count[i] > max_count) {
+                max_count = vote_count[i];
+                majority_label = i;
+            }
+        }
+        // Assign the majority label to the query point
+        query_labels[query_idx] = majority_label;
+    }
+
+}
+
+
+bool knn_cuda_global(const float * ref,
+                     int           ref_nb,
+                     const float * query,
+                     int           query_nb,
+                     int           dim,
+                     int           k,
+                     float *       knn_dist,
+                     int *         knn_index, 
+                     int *         query_labels, double &elapsed_time) {
+
+    // Constants
+    const unsigned int size_of_float = sizeof(float);
+    const unsigned int size_of_int   = sizeof(int);
+
+    // Return variables
+    cudaError_t err0, err1, err2, err3, err4;
+
+    // Check that we have at least one CUDA device 
+    int nb_devices;
+    err0 = cudaGetDeviceCount(&nb_devices);
+    if (err0 != cudaSuccess || nb_devices == 0) {
+        printf("ERROR: No CUDA device found\n");
+        return false;
+    }
+
+    // Select the first CUDA device as default
+    err0 = cudaSetDevice(0);
+    if (err0 != cudaSuccess) {
+        printf("ERROR: Cannot set the chosen CUDA device\n");
+        return false;
+    }
+
+    // Allocate global memory
+    float * ref_dev   = NULL;
+    float * query_dev = NULL;
+    float * dist_dev  = NULL;
+    int   * index_dev = NULL;
+    int   * query_labels_dev = NULL;
+    size_t  ref_pitch_in_bytes;
+    size_t  query_pitch_in_bytes;
+    size_t  dist_pitch_in_bytes;
+    size_t  index_pitch_in_bytes;
+    printf("Attempting to allocate memory...\n");
+    err0 = cudaMallocPitch((void**)&ref_dev,   &ref_pitch_in_bytes,   ref_nb   * size_of_float, dim);
+    err1 = cudaMallocPitch((void**)&query_dev, &query_pitch_in_bytes, query_nb * size_of_float, dim);
+    err2 = cudaMallocPitch((void**)&dist_dev,  &dist_pitch_in_bytes,  query_nb * size_of_float, ref_nb);
+    err3 = cudaMallocPitch((void**)&index_dev, &index_pitch_in_bytes, query_nb * size_of_int,   k);
+    err4 = cudaMalloc((void**)&query_labels_dev, query_nb * size_of_int);
+
+    if (err0 != cudaSuccess) {
+        printf("ERROR: Memory allocation error for ref_dev: %s\n", cudaGetErrorString(err0));
+    }
+    if (err1 != cudaSuccess) {
+        printf("ERROR: Memory allocation error for query_dev: %s\n", cudaGetErrorString(err1));
+    }
+    if (err2 != cudaSuccess) {
+        printf("ERROR: Memory allocation error for dist_dev: %s\n", cudaGetErrorString(err2));
+    }
+    if (err3 != cudaSuccess) {
+        printf("ERROR: Memory allocation error for index_dev: %s\n", cudaGetErrorString(err3));
+    }
+    if (err4 != cudaSuccess) {
+        printf("ERROR: Memory allocation error for query_labels_dev: %s\n", cudaGetErrorString(err4));
+    }
```

Comment:
if you check this after each call you would not need to declare different variable for this. Also if you exit after an error occur following if condition won't be necessary.